kresd: runtime fixes

eadwu opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Not sure about the directory permissions.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

eadwu committed:
nixos/kresd: ensure /var/lib/knot-resolver exists

eadwu committed:
nixos/kresd: ensure /run/knot-resolver exists

vcunat commented:
Thanks for ping.  I'm not sure I understand &ndash; have you ran into some issues?  There's tmpfiles spec that should ensure existence of all three.  The one for cache was added here as well to
>  \# Ensure correct ownership in case UID or GID changes.

because of ae74a0e27.  The other two didn't exist before that commit, so they weren't added in the same way.

vcunat commented:
To be clear, I see no harm in adding this, so I'm mainly curious.  (and perhaps the comments are a bit confusing in combination with these "invisible" tmpfiles)

eadwu commented:
For some reason the files weren't created on my system. Kept getting CHDIR errors for (`/var/lib/knot-resolver`) and then errors about `/run/knot-resolver` not existing for the control.

vcunat commented:
Oh, and `lib/systemd/` does get applied?  (from `systemd.packages = [ package ];`)

flokli commented:
https://github.com/NixOS/nixpkgs/pull/95565 was merged, which should fix the `knot-resolver` tmpfiles not being applied.

Even on master, the permissions in the `kresd@` template unit differ with what's in the `tmpfiles` file. `tmpfiles` will be applied once on boot, and on each system activation, whereas systemd will apply `*DirectoryMode` when starting the service, too (thus shadowing the tmpfiles).

@vcunat I think we can use only the per-unit directives here. I sent a patch to upstream at knot-resolver@labs.nic.cz - not sure how to exacty contribute. Also opened https://gitlab.nic.cz/knot/knot-resolver/-/issues/596.

flokli commented on nixos/modules/services/networking/kresd.nix:
```diff
@@ -129,12 +129,18 @@ in {
         + "-c ${package}/lib/knot-resolver/distro-preconfig.lua -c ${configFile}";
+      # Ensure /run/knot-resolver exists
+      RuntimeDirectory = "knot-resolver";
+      RuntimeDirectoryMode = "0750";
+      # Ensure /var/lib/knot-resolver exists
+      StateDirectory = "knot-resolver";
+      StateDirectoryMode = "0750";
       # Ensure correct ownership in case UID or GID changes.
       CacheDirectory = "knot-resolver";
       CacheDirectoryMode = "0750";
```
These permissions should all be 0770 (to be consistent with what's in the tmpfiles file)

eadwu replied:
Done.

vcunat commented:
Yes, I think on NixOS the kresd service can abandon using tmpfiles in favor of these, though I can't really see any practical differences.

eadwu committed:
nixos/kresd: fix CacheDirectory permissions as per tmpfiles

eadwu committed:
nixos/kresd: remove derivation from systemd.tmpfiles

Using per-unit directives as per https://github.com/NixOS/nixpkgs/pull/95222#issuecomment-674512571
